diffrax/_brownian/tree.py

```diff
@@ -369,14 +429,114 @@ def _body_fun(_state: _State):
         final_state = lax.while_loop(_cond_fun, _body_fun, init_state)
 
         s = final_state.s
-        su = 2.0**-final_state.level
+        su = jnp.asarray(2.0**-final_state.level, dtype=tdtype)
 
         sr = jax.nn.relu(r - s)
         # make sure su = sr + ru regardless of cancellation error
         ru = jax.nn.relu(su - sr)
 
         w_s, w_u, w_su = final_state.w_s_u_su
-        if self.levy_area is SpaceTimeLevyArea:
+
+        if self.levy_area is SpaceTimeTimeLevyArea:
+            # Based on Theorem 3.7 from the paper on
+            # Single-seed generation of Brownian paths and integrals
+            assert final_state.bhh_s_u_su is not None
+            assert final_state.bkk_s_u_su is not None
+            bhh_s, bhh_u, bhh_su = final_state.bhh_s_u_su
+            bkk_s, bkk_u, bkk_su = final_state.bkk_s_u_su
+
+            su3 = jnp.power(su, 3)
+            sr_by_su = sr / su
+            sr_by_su_3 = jnp.power(sr_by_su, 3)
+            sr_by_su_5 = jnp.power(sr_by_su, 5)
+            ru_by_su = ru / su
+            sr_ru_by_su2 = sr_by_su * ru_by_su
+            sr2 = jnp.square(sr)
+            ru2 = jnp.square(ru)
+            su2 = jnp.square(su)
+
+            # compute the mean of (W_sr, H_sr, K_sr) conditioned on
+            # (W_s, H_s, K_s, W_u, H_u, K_u)
+            with jax.numpy_dtype_promotion("standard"):
+                bb_mean = (6 * sr_ru_by_su2 / su) * bhh_su + (
+                    120 * sr_ru_by_su2 * (0.5 - sr_by_su) / su2
+                ) * bkk_su
+                w_mean = sr_by_su * w_su + bb_mean
+                h_mean = (sr_by_su**2 / su) * (bhh_su + (30 * ru_by_su / su) * bkk_su)
+                k_mean = (sr_by_su_3 / su2) * bkk_su
+
+                # compute the covariance matrix of (W_sr, H_sr, K_sr) conditioned on
+                # (W_s, H_s, K_s, W_u, H_u, K_u)
+                ww_cov = (
+                    sr_by_su * ru_by_su * ((sr - ru) ** 4 + 4 * (sr2 * ru2))
+                ) / su3
+                wh_cov = -(sr_by_su_3 * ru_by_su * (sr2 - 3 * sr * ru + 6 * ru2)) / (
+                    2 * su
+                )
+                wk_cov = (sr_by_su**4) * ru_by_su * (sr - ru) / 12
+                hh_cov = (
+                    sr / 12 * (1 - sr_by_su_3 * (sr2 + 2 * sr * ru + 16 * ru2) / su2)
+                )
+                hk_cov = -(ru / 24) * sr_by_su_5
+                kk_cov = (sr / 720) * (1.0 - sr_by_su_5)
+
+            cov = jnp.array(
+                [
+                    [ww_cov, wh_cov, wk_cov],
+                    [wh_cov, hh_cov, hk_cov],
+                    [wk_cov, hk_cov, kk_cov],
+                ]
+            )
+
+            if self._spline == "sqrt":
+                # NOTE: not compatible with jnp.float16
```

Comment:
I think we should still aim to handle this. Probably this can be done by casting the dtype?
```python
dtype_atleast32 = jnp.result_type(dtype, jnp.float32)
hat_y = jr.multivariate_normal(..., dtype=dtype_atleast32)
hat_y = hat_y.astype(dtype)
```

Reply:
FWIW I think generating a multivariate normal shouldn't be too tricky if we wanted to just do it manually.

I also don't love that this is apparently calling out to SVD. Admittedly, though, I have just done a quick google for an explicit form for the square root of a 3x3 posdef matrix and it looks like there isn't one...

Reply:
I did the Cholesky decomp manually, but it's very complicated and in my very basic experiments it took slightly longer to compute it with that formula (so computing the decomposition directly from s, r, u), as opposed to first computing the cov matrix, and then doing SVD on that. The issue with Cholesky is that it is very imprecise when the matrix is close to singular (which happens when r is close to s or to u). SVD is the only good option for near-singular matrices. Alternatively we'd have to do a separate case for when r-s or u-r is small, which is probably not ideal.

Reply:
Yes, the type-casting idea is very good.